lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -2182,8 +2182,14 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
           DbgTy.getAlignment(), DbgTy.hasDefaultAlignment(),
           /* IsMetadataType = */ false, DbgTy.isFixedBuffer(),
           DbgTy.getNumExtraInhabitants());
-      return DBuilder.createTypedef(getOrCreateType(AliasedDbgTy), MangledName,
-                                    L.File, 0, Scope);
+      auto *TypeDef = DBuilder.createTypedef(getOrCreateType(AliasedDbgTy),
+                                             MangledName, L.File, 0, Scope);
+      // Bound generic types don't reference their type parameters in ASTTypes
+      // mode, so we need to artificially keep typealiases alive, since they can
+      // appear in reflection metadata.
+      if (Opts.DebugInfoLevel < IRGenDebugInfoLevel::DwarfTypes)
+        DBuilder.retainType(TypeDef);
```

Comment:
I thought `retainType` would retain the type in the IR but not in DWARF, am I wrong?

Reply:
It does both:
https://github.com/llvm/llvm-project/blob/5d6d982df61d16b6d498e6d59dd91c059679d3d8/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#L1221

Reply:
Hmm, does dsymutil keep them too? I think maybe that's what I was remembering. LGTM if yes

Reply:
It should not, that's a great point.

Reply:
Here is a better solution that builds on top of https://github.com/swiftlang/swift/pull/78965.